### PR TITLE
8287292: Improve TransformKey to pack more kinds of transforms efficiently

### PR DESCRIPTION
**Pull request by cl4es**
The bespoke caching scheme in `jl.invoke.LambdaFormEditor.TransformKey` allows keys to be compacted when all byte values of the key fit in 4 bits, otherwise a byte array is allocated and used. This means that all transforms with a kind value above 15 will be forced to allocate and use array comparisons.

Removing unused and folding some transforms to ensure all existing kinds can fit snugly within the 0-15 value range realize a minor improvement to footprint, speed and allocation pressure of affected transforms, e.g. ~300bytes/op reduction in the `StringConcatFactoryBootstraps` microbenchmark:

Baseline:
```
Benchmark                                                      Mode  Cnt     Score     Error   Units
SCFB.makeConcatWithConstants                                   avgt   15  2048.475 ?  69.887   ns/op
SCFB.makeConcatWithConstants:?gc.alloc.rate.norm               avgt   15  3487.311 ?  80.385    B/op
```

Patched:
```
Benchmark                                                      Mode  Cnt     Score     Error   Units
SCFB.makeConcatWithConstants                                   avgt   15  1961.985 ? 101.519   ns/op
SCFB.makeConcatWithConstants:?gc.alloc.rate.norm               avgt   15  3156.478 ? 183.600    B/op
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287292](https://bugs.openjdk.java.net/browse/JDK-8287292): Improve TransformKey to pack more kinds of transforms efficiently

### Reviewers
 * [Jim Laskey](https://openjdk.java.net/census#jlaskey) (@JimLaskey - **Reviewer**) ⚠️ Review applies to [001e9d16](https://git.openjdk.java.net/jdk/pull/8881/files/001e9d1671fd3b19cf9cc0bf306fe5c062613a3c)
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [2be3b25c](https://git.openjdk.java.net/jdk/pull/8881/files/2be3b25cd95f6b5e14bbff8240559b5d244b54da)
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [612b4ece](https://git.openjdk.java.net/jdk/pull/8881/files/612b4eceb891008e0475e08116e230a49f36d6e7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8881/head:pull/8881` \
`$ git checkout pull/8881`

Update a local copy of the PR: \
`$ git checkout pull/8881` \
`$ git pull https://git.openjdk.java.net/jdk pull/8881/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8881`

View PR using the GUI difftool: \
`$ git pr show -t 8881`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8881.diff">https://git.openjdk.java.net/jdk/pull/8881.diff</a>

</details>

### CODE REVIEW

**Review by JimLaskey — APPROVED** [review]

File: src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java [hunk 1]
```diff
@@ -231,6 +229,22 @@ private static TransformKey ofBothArrays(byte kind, int[] b123, byte[] b456) {
...
                 PACKED_BYTE_MAX_LENGTH = (STRESS_TEST ? 3 : 64 / PACKED_BYTE_SIZE);
 
+        private static long packedBytes(byte b0, int b1, int[] b23456) {
+            if (b23456.length + 2 > PACKED_BYTE_MAX_LENGTH)
+                return 0;
+            long pb = 0;
+            int bitset = b0 | b1;
+            for (int i = 0; i < b23456.length; i++) {
+                int b = b23456[i] & 0xFF;
+                bitset |= b;
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:239]: Looks like `b` is always truncated. I wonder what happens if the ints in this array are larger than a byte (which seems to be possible in e.g. the case of argument positions). Some higher order bits might be dropped, but the resulting `b` might only have the least significant 4 bits set.

I think the untruncated value should be used to compute the bitset? `butset |= b23456[i]`?

Comment (JornVernee) [reply src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:239]: Maybe not... argument positions should fit in a byte as well. But, maybe there are other problematic cases? Or are the ints guaranteed to fit in a byte?

Comment (JornVernee) [reply src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:239]: Yes, please. IMHO it makes it clear that that is the assumption of this code.

**Review by JornVernee — APPROVED** [review]

File: src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java [hunk 2]
```diff
@@ -231,6 +237,52 @@ private static TransformKey ofBothArrays(byte kind, int[] b123, byte[] b456) {
...
+            long pb = 0;
+            int bitset = b0 | b1 | b2;
+            for (int i = 0; i < b345.length; i++) {
+                int b = ival(b345[i]);
+                bitset |= b;
+                pb |= (long)b << ((i + 3) * PACKED_BYTE_SIZE);
+            }
+            if (!inRange(bitset))
+                return 0;
+            pb = pb | b2 << (2 * PACKED_BYTE_SIZE) | b1 << PACKED_BYTE_SIZE | b0;
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:252]:
```suggestion
            pb = pb | packagedBytes(b0, b1, b2);
```

File: src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java [hunk 3]
```diff
@@ -186,43 +191,44 @@ static TransformKey of(byte b0, int b1, int b2, int b3) {
...
-            return ofBothArrays(kind, new int[]{ b1, b2 }, b345);
+            return new TransformKey(fullBytes);
         }
-        private static TransformKey ofBothArrays(byte kind, int[] b123, byte[] b456) {
-            byte[] fullBytes = new byte[1 + b123.length + b456.length];
-            int i = 0;
-            fullBytes[i++] = bval(kind);
-            for (int bv : b123) {
-                fullBytes[i++] = bval(bv);
+        static TransformKey of(byte kind, int b1, int b2, int[] b3456) {
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:219]: Nit: I can't figure out if `b3456` intends to be a different convention than the others `b123` and `b234`.  I would expect something like this:

```suggestion
        static TransformKey of(byte kind, int b1, int b2, int... b345) {
```

File: src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java [hunk 4]
```diff
@@ -231,6 +237,52 @@ private static TransformKey ofBothArrays(byte kind, int[] b123, byte[] b456) {
...
+            long pb = 0;
+            int bitset = b0 | b1;
+            for (int i = 0; i < b234.length; i++) {
+                int b = ival(b234[i]);
+                bitset |= b;
+                pb |= (long)b << ((i + 2) * PACKED_BYTE_SIZE);
+            }
+            if (!inRange(bitset))
+                return 0;
+            pb = pb | b1 << PACKED_BYTE_SIZE | b0;
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/LambdaFormEditor.java:267]:
```suggestion
            pb = pb | packagedBytes(b0, b1);
```

**Review by mlchung — APPROVED** [review]
LGTM with a couple of suggestions.

**Comment by cl4es** [comment]
Thanks!

/integrate

**Pull request closed without merging**